### PR TITLE
added Exception handling for devices

### PR DESCRIPTION
**Pull request by ZohebShaikh**
closes https://github.com/DiamondLightSource/blueapi/issues/444

The main change in this PR is that `make_all_devices` returns tuple of `(all_devices, exception_devices)` 

### Instructions to reviewer on how to test:
1. Checkout branch `device-resilience` in dodal
2. In a common python virtual environment install both blueapi and dodal 
`pip install -e ../dodal[dev]`  `pip install -e .[dev]`
3. Test device does not fail when exception is occurred

### Checks for reviewer
- [ ] Achives requirement for  https://github.com/DiamondLightSource/blueapi/issues/444

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -66,7 +66,9 @@ def test_no_signature_builtins_not_devices() -> None:
     import tests.fake_beamline_misbehaving_builtins as fake_beamline
 
     devices = make_all_devices(fake_beamline)
-    assert not devices
+    assert not devices[0]
+    for exception in devices[1]:
```
Comment (ZohebShaikh) [inline tests/test_utils.py:70]: I just had a look at this, the device fake_beamlines don't create exceptions so I will need to add some devices in the fake_beamline that raises a exception when it does not connect

File: tests/beamlines/unit_tests/test_beamline_utils.py [hunk 2]
```diff
@@ -69,7 +69,7 @@ def test_instantiate_v2_function_fake_makes_fake():
 
 def test_clear_devices(RE):
     devices = make_all_devices(i03, fake_with_ophyd_sim=True)
-    assert len(beamline_utils.ACTIVE_DEVICES) == len(devices.keys())
+    assert len(beamline_utils.ACTIVE_DEVICES) == len(devices[0].keys())
```
Comment (callumforrester) [inline tests/beamlines/unit_tests/test_beamline_utils.py:72]: Nit: You could instead have

```python
devices, exceptions = make_all_devices(i03, fake_with_ophyd_sim=True)
assert len(beamline_utils.ACTIVE_DEVICES) == len(devices.keys()) and len(exceptions) == 0
```

...which is more readable, ditto for the other tests

File: src/dodal/utils.py [hunk 3]
```diff
@@ -109,7 +110,7 @@ def wrapper(*args, **kwds) -> T:
 
 def make_all_devices(
```
Comment (callumforrester) [inline src/dodal/utils.py:111]: Should: The docstring for this function needs updating

File: tests/test_utils.py (hunk 1 above)
Comment (callumforrester) [reply tests/test_utils.py:70]: I think we need the following test cases:
- All devices pass
- All devices fail
- Mixture
If you have to make a few small fake beamline modules to cover those then go for it

File: src/dodal/utils.py [hunk 4]
```diff
@@ -125,16 +126,19 @@ def make_all_devices(
...
     return devices
 
 
 def invoke_factories(
     factories: Mapping[str, AnyDeviceFactory],
     **kwargs,
-) -> Dict[str, AnyDevice]:
+) -> Tuple[Dict[str, AnyDevice], Dict[str, Exception]]:
     devices: dict[str, AnyDevice] = {}
+    exception_devices: dict[str, Exception] = {}
```
Comment (callumforrester) [inline src/dodal/utils.py:141]: Nit: I think this makes more sense

```suggestion
    device_exceptions: dict[str, Exception] = {}
```

**Review by callumforrester — CHANGES_REQUESTED** [review]
LGTM! Just the tests and a couple of nitpicks

File: src/dodal/utils.py [hunk 5]
```diff
@@ -150,11 +158,18 @@ def invoke_factories(
...
-    all_devices = {device.name: device for device in devices.values()}
+        try:
+            devices[dependent_name] = factories[dependent_name](**params, **kwargs)
+        except Exception as e:
+            device_exceptions[dependent_name] = e
+
+    all_devices = {
+        device.name: device
+        for device in devices.values()
+        if not isinstance(device, Exception)
```
Comment (callumforrester) [inline src/dodal/utils.py:169]: I don't think you need this line

**Comment by callumforrester** [comment]
I'm taking over this PR in @ZohebShaikh's absence

File: tests/fake_beamline_all_devices_rasie_exception.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+def device_a() -> Readable:
+    raise DestroyedError
+
+
+def device_b() -> EpicsMotor:
+    raise DisconnectedError
+
+
+def device_c() -> Cryo:
+    raise UnknownStatusFailure
```
Comment (DiamondJoseph) [inline tests/fake_beamline_all_devices_rasie_exception.py:21]: Currently one Protocol, 2 ophyd devices, add an ophyd-async device class?

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]
Overall looks good just some edge cases and minor optimisations

File: tests/fake_beamline_all_devices_rasie_exception.py (hunk 6 above)
Comment (callumforrester) [reply tests/fake_beamline_all_devices_rasie_exception.py:21]: Given that we are moving in the direction of exclusive ophyd-async (and therefore hopefully dodal's handling of ophyd devices can go away one day) I might just make it ophyd-async exclusively

Comment (DiamondJoseph) [reply tests/fake_beamline_all_devices_rasie_exception.py]: I'd prefer to see both tested until there are no ophyd devices, given they are different execution paths

Comment (callumforrester) [reply tests/fake_beamline_all_devices_rasie_exception.py]: Have reverted cryostream

File: src/dodal/utils.py [hunk 7]
```diff
@@ -120,41 +121,49 @@ def make_all_devices(
...
+    )
 
     return devices
 
 
 def invoke_factories(
     factories: Mapping[str, AnyDeviceFactory],
     **kwargs,
-) -> Dict[str, AnyDevice]:
+) -> Tuple[Dict[str, AnyDevice], Dict[str, Exception]]:
```
Comment (DominicOram) [inline src/dodal/utils.py:142]: Could: I think it's worth a bit of a docstring here on how the two dicts relate to each other. e.g. that every device will be represented in one list, either the first if it succeeds or the second if it fails? I wonder if a type like `Dict[str, AnyDevice | Exception]` would make this more clear but I'm not too sure.

Comment (DominicOram) [reply src/dodal/utils.py:142]: I also thing a general `docstring` on what's going on with the leaves etc. would be nice

**Review by DominicOram — CHANGES_REQUESTED** [review]
Ta, I think mostly just need to fix the `cli`

File: src/dodal/utils.py (hunk 7 above)
Comment (callumforrester) [reply src/dodal/utils.py:142]: Have added some docstrings, the other option which I'm tempted by is to return a dataclass rather than a tuple. Will do that if no one objects.

Comment (DominicOram) [reply src/dodal/utils.py:142]: Fine by me

**Review by DominicOram — APPROVED** [review]
Great, thank you, happy for this to be merged now but if you want to do the `dataclass` change then that's fine too

**Review by DiamondJoseph — APPROVED** [review]
